Source/WebCore/dom/ContentVisibilityDocumentState.cpp

```diff
@@ -0,0 +1,176 @@
+
+void ContentVisibilityDocumentState::unobserve(Element& element)
+{
+    auto& state = element.document().contentVisibilityDocumentState();
+    if (auto& intersectionObserver = state.m_observer) {
+        intersectionObserver->unobserve(element);
+        state.updateOnScreenObservationTarget(element, false);
+    }
+    element.setContentRelevancyStatus({ });
+}
+
+IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Document& document)
+{
+    if (!m_observer) {
+        auto callback = ContentVisibilityIntersectionObserverCallback::create(document);
+        IntersectionObserver::Init options { &document, { }, { } };
+        auto observer = IntersectionObserver::create(document, WTFMove(callback), WTFMove(options));
+        if (observer.hasException())
+            return nullptr;
+        m_observer = observer.returnValue().ptr();
+    }
+    return m_observer.get();
+}
+
+bool ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancyStatus>& relevancyToCheck)
+{
+    bool didUpdateAnyContentRelevancy = false;
+    for (auto target : m_observer->observationTargets()) {
+        if (target) {
+            auto oldRelevancy = target->contentRelevancyStatus();
+            auto newRelevancy = oldRelevancy;
+            auto setRelevancyValue = [&](ContentRelevancyStatus reason, bool value) {
+                if (value)
+                    newRelevancy.add(reason);
+                else
+                    newRelevancy.remove(reason);
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::OnScreen))
+                setRelevancyValue(ContentRelevancyStatus::OnScreen, m_onScreenObservationTargets.contains(*target));
+
+            if (relevancyToCheck.contains(ContentRelevancyStatus::Focused))
+                setRelevancyValue(ContentRelevancyStatus::Focused, target->hasFocusWithin());
+
+            auto hasTopLayerinSubtree = [](const Element& target) {
+                for (auto& element : target.document().topLayerElements()) {
+                    if (element->isDescendantOf(target))
+                        return true;
+                }
+                return false;
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::IsInTopLayer))
+                setRelevancyValue(ContentRelevancyStatus::IsInTopLayer, hasTopLayerinSubtree(*target));
+
+            if (oldRelevancy == newRelevancy)
+                continue;
+            target->setContentRelevancyStatus(newRelevancy);
+            target->invalidateStyle();
+            didUpdateAnyContentRelevancy = true;
+        }
+    }
+    return didUpdateAnyContentRelevancy;
+}
+
+// Workaround for lack of support for scroll anchoring. We make sure any content-visibility: auto elements
+// above the one to be scrolled to are already hidden, so the scroll position will not need to be adjusted
+// later.
+// FIXME: remove when scroll anchoring is implemented.
```

Comment:
This needs a bug filed that's related to the scroll anchoring bug so we don't forget to remove this workaround.

Reply:
Good point, I opened https://bugs.webkit.org/show_bug.cgi?id=259269 for this.